datalad/distributed/drop.py

```diff
@@ -415,7 +415,9 @@ def _drop_dataset(ds, paths, what, reckless, recursive, recursion_limit, jobs):
                 res['message'] = message
             error_messages = r.get('error-messages')
```

Comment:
TL;DR: I think we would benefit from some `AnnexRepo.annex_to_datalad_error` `@staticmethod` helper or alike to be reused across all those uses.

uff... venting about my quality time to understand what we are talking about. We would benefit from typing the result record into some `@dataclass` and adding type checking... First I have spent time hunting for `error-message` (not `error_message`) to find none!

<details>
<summary>I did find "error-messages"</summary> 

```shell
$> git grep -p "[\"']error-messages"   
CHANGELOG.md=only supported Python flavor.
CHANGELOG.md:  relay information from the "error-messages" field.  ([#3931][])
datalad/distributed/drop.py=def _drop_dataset(ds, paths, what, reckless, recursive, recursion_limit, jobs):
datalad/distributed/drop.py:            error_messages = r.get('error-messages')
datalad/distributed/tests/test_ora_http.py=def test_read_access(store_path, store_url, ds_path):
datalad/distributed/tests/test_ora_http.py:        assert_equal(r["error-messages"], [])
datalad/interface/results.py=def annexjson2result(d, ds, **kwargs):
datalad/interface/results.py:    if d.get('error-messages', None):
datalad/interface/results.py:        res['error_message'] = '\n'.join(m.strip() for m in d['error-messages'])
datalad/local/copy_file.py=def _place_filekey(finfo, str_src, dest, str_dest, dest_repo):
datalad/local/copy_file.py:                m for r in res for m in r.get('error-messages', [])),
datalad/runner/exception.py=def _format_json_error_messages(recs: List[Dict]):
datalad/runner/exception.py:            '\n'.join(r.get('error-messages', [])),
datalad/runner/tests/test_exception.py=def get_json_objects(object_count, message_count) -> List[Dict]:
datalad/runner/tests/test_exception.py:            "error-messages": [
datalad/support/annex_utils.py=def _fake_json_for_non_existing(paths, cmd):
datalad/support/annex_utils.py:             "error-messages": ["File unknown to git"]  # Note,
datalad/support/annexrepo.py=class AnnexRepo(GitRepo, RepoInterface):
datalad/support/annexrepo.py:                       for k in j if k != 'file' and k != 'error-messages'})
datalad/support/annexrepo.py:            if j.get('error-messages', None):
datalad/support/annexrepo.py:                rec['error_message'] = '\n'.join(m.strip() for m in j['error-messages'])
datalad/support/annexrepo.py:                    message='\n'.join(r['error-messages'])
datalad/support/annexrepo.py:                    if 'error-messages' in r else None,
datalad/support/annexrepo.py=class AnnexJsonProtocol(WitlessProtocol):
datalad/support/annexrepo.py:            for err_msg in action.pop('error-messages', []):
datalad/support/gitrepo.py=class GitRepo(CoreGitRepo):
datalad/support/gitrepo.py:                    message='\n'.join(r['error-messages'])
datalad/support/gitrepo.py:                    if 'error-messages' in r else None,
datalad/support/tests/test_annexrepo.py=def test_annex_drop(src, dst):
datalad/support/tests/test_annexrepo.py:    # CommandError has to pull the errors from the JSON record 'error-messages'
datalad/support/tests/test_annexrepo.py=def test_error_reporting(path):
datalad/support/tests/test_annexrepo.py:            'error-messages': ['File unknown to git'],
datalad/support/tests/test_annexrepo.py=def test_save_noperms(path):
datalad/support/tests/test_annexrepo.py:        assert_in('permission denied', res[0]['error-messages'][0])
```
</details>

confused, and re-reading spotted the difference in `-` vs `_`, so indeed there is `error_message` but *really just few*:

```
$> git grep -p "[\"']error_message[^s]"
datalad/interface/results.py=def get_status_dict(action=None, ds=None, path=None, type=None, logger=None,
datalad/interface/results.py:        d['error_message'] = error_message
datalad/interface/results.py=def annexjson2result(d, ds, **kwargs):
datalad/interface/results.py:        res['error_message'] = '\n'.join(m.strip() for m in d['error-messages'])
datalad/interface/utils.py=def generic_result_renderer(res):
datalad/interface/utils.py:                res['error_message'][0] % res['error_message'][1:]
datalad/interface/utils.py:                if isinstance(res['error_message'], tuple) else res[
datalad/interface/utils.py:                    'error_message']), ac.RED)
datalad/interface/utils.py:            if res.get('error_message', None) and res.get('status', None) != 'ok' else ''))
datalad/support/annexrepo.py=class AnnexRepo(GitRepo, RepoInterface):
datalad/support/annexrepo.py:                rec['error_message'] = '\n'.join(m.strip() for m in j['error-messages'])
```

and only then, looking at above duplications of conversions from `error-messages` to  `error_message` that we are talking about duplicated handling of `error-messages` we receive from direct invocations of git-annex (now spread out through code base instead of handled uniformly within `AnnexRepo`.

Reply:
Thanks for taking a closer look! Sorry if I didn't make it clearer about the spelling.

Let me just add that `error_messages` (underscore / plural) that started this issue, and this PR changes into `error_message` does not exist outside `drop` at all (below I'm looking at master, but usage in annexrepo has been since removed in maint), therefore I decided to change `drop` and not the renderer:

```
❱ git grep -p "[^_format_json_]error_messages"
datalad/distributed/drop.py=def _drop_dataset(ds, paths, what, reckless, recursive, recursion_limit, jobs):
datalad/distributed/drop.py:            error_messages = r.get('error-messages')
datalad/distributed/drop.py:            if error_messages:
datalad/distributed/drop.py:                res['error_messages'] = error_messages
datalad/support/annexrepo.py=class AnnexRepo(GitRepo, RepoInterface):
datalad/support/annexrepo.py:                       for k in j if k != 'file' and k != 'error_messages'})
datalad/support/annexrepo.py:            # change annex' `error_messages` into singular to match result
```

> I think we would benefit from some AnnexRepo.annex_to_datalad_error @staticmethod helper or alike to be reused across all those uses

There is, in a way, `annexjson2result` in `datalad/interface/results.py`, which maybe could be reused?